Test/tutorial/induction-principle.md

```diff
@@ -0,0 +1,782 @@
+}
+```
+
+As expected, only the `Op` case of the induction principle requires manual work, by calling
+the lemma [`Pure.InterpExpr_Pure`](#interpexpr-pure) that we proved above, which states that
+evaluating an expression which satisfies the `IsPure` predicate leaves the state unchanged.
+
+## Verifying `VarUnchanged`
+
+We also applied the induction principle to the proof that if an expression syntactically
+doesn't update a free variable (i.e., not bound by a `let`), then this variable is not
+updated during the evaluation.
+
+We define a predicate `VarUnchanged(x: string, e: Expr)`, which returns `true` if the variable
+`x` is not updated in expression `e`, while ignoring the places where it is shadowed by a let-binding.
+The interesting cases are `Bind(bvar: string, bval: Expr, body: Expr)` and
+`Assign(avar: string, aval: Expr)`.
+
+```dafny
+predicate method VarUnchanged(x: string, e: Expr)
+  // Returns true if no assignments of `x` (not shadowed by a let-binding) appears
+  // in `e`.
+{
+  match e
+    case Bind(bvar, bval, body) =>
+      // The rhs doesn't update x
+      VarUnchanged(x, bval) &&
+      // If the binding doesn't shadow x, the body of the let-binding doesn't update x
+      (bvar == x || VarUnchanged(x, body))
+
+    case Assign(avar, aval) =>
+      // The variable we update is not x
+      avar != x
+      // And we don't update x in the rhs
+      && VarUnchanged(x, aval)
+
+    // Other cases omitted
+    case ... =>
+      ...
+}
+
+```
+
+The property we want to prove is the following one:
+```dafny
+forall x, ctx, e ::
+  VarUnchanged(x, e) ==>
+  x in ctx ==>
+  match InterpExpr(e, st.ctx)
+    case Success((_, ctx')) =>
+      && x in ctx'.Keys // x is also in the new environment
+      && ctx'[x] == ctx[x] // x maps to the same value as before
+    case Failure =>
+      true
+  Pre(st, e) ==> ResultSameX(st, res)
+```
+
+Because of the potential shadowing appearing in the `Bind` case, we actually can't use this
+property directly in the inductive proof. We actually have to make the `x` string optional
+(see the [code development](induction-principle/VarUnchanged.dfy) for the full details). Once this detail properly handled, all the
+proofs go through automatically.
+
+
+<a id="section-increments"></a>
+
+## Incrementally expanding the language
+
+We now study how incremental updates to the AST impact the induction
+principle and the proofs we have done so far.
+
+<a id="section-increments-seq"></a>
+
+### A more general version of `Seq`: `Seq(expr, expr) ---> Seq(seq<Expr>)`
+
+The first modification we do is to transform the `Seq` case from `Seq(e1: Expr, e2: Expr)`
+to `Seq(seq<Expr>)`. This requires adding a function to interpret sequences of expressions.
+Anticipating a bit on the next language extension we will implement (allowing assignments
+and let-bindings to update/declare a sequence of variables -
+see [next section](#section-increments-varseq)), we define this function so that it returns
```

Comment:
```suggestion
Anticipating a bit on the next language extension we will implement allowing assignments
and let-bindings to update/declare a sequence of variables -
see [next section](#section-increments-varseq); we define this function so that it returns
```